framework/src/org/checkerframework/framework/type/DefaultRawnessComparer.java

```diff
@@ -253,50 +269,55 @@ public Boolean visitWildcardSupertype(
             return true;
         }
 
-        if (checkOrAdd(subtype, superExtendsBound, visited)) {
+        if (visited.contains(subtype, superExtendsBound, currentTop)) {
             return true;
         }
 
-        return this.visit(subtype, superExtendsBound, visited);
+        Boolean result = this.visit(subtype, superExtendsBound, visited);
+        visited.add(subtype, supertype, currentTop, result);
+        visited.add(subtype, superExtendsBound, currentTop, result);
+        return result;
     }
 
     public Boolean visitTypevarSupertype(
             AnnotatedTypeMirror subtype, AnnotatedTypeVariable supertype, VisitHistory visited) {
-        if (checkOrAdd(subtype, supertype, visited)) {
+        if (visited.contains(subtype, supertype, currentTop)) {
             return true;
         }
 
         final AnnotatedTypeMirror supertypeUb = supertype.getUpperBound();
-        if (checkOrAdd(subtype, supertypeUb, visited)) {
+        if (visited.contains(subtype, supertypeUb, currentTop)) {
             return true;
         }
 
-        return this.visit(subtype, supertypeUb, visited);
+        Boolean result = this.visit(subtype, supertypeUb, visited);
+        visited.add(subtype, supertype, currentTop, result);
+        visited.add(subtype, supertypeUb, currentTop, result);
+        return result;
     }
 
     public Boolean visitTypeVarSubtype(
             AnnotatedTypeVariable subtype, AnnotatedTypeMirror supertype, VisitHistory visited) {
-        if (checkOrAdd(subtype, supertype, visited)) {
+        if (visited.contains(subtype, supertype, currentTop)) {
             return true;
         }
 
         final AnnotatedTypeMirror subtypeUb = subtype.getUpperBound();
-        if (checkOrAdd(subtypeUb, supertype, visited)) {
+        if (visited.contains(subtypeUb, supertype, currentTop)) {
             return true;
         }
 
-        return this.visit(subtype, subtypeUb, visited);
-    }
-
-    private boolean checkOrAdd(
-            final AnnotatedTypeMirror subtype,
-            final AnnotatedTypeMirror supertype,
-            final VisitHistory visited) {
-        if (visited.contains(subtype, supertype)) {
-            return true;
-        }
-
-        visited.add(subtype, supertype);
-        return false;
+        // Needed to prevent infinite recursion e.g. in
+        // tests/all-systems/java8inference/Issue1815.java
+        visited.add(subtype, supertype, currentTop, true);
+
+        // TODO: tests/all-systems/InferTypeArgsCondtionalExpression.java
+        // breaks with this:
+        //Boolean result = this.visit(subtypeUb, supertype, visited);
+        // it doesn't make sense to me to compare subtype and subtypeUB
+        Boolean result = this.visit(subtype, subtypeUb, visited);
```

Comment:
Highlighting this comment and the test that fails when using a more meaningful check.